### ISSUE TITLE
[DOC] How to simply remap keys, including diacritics/deadkeys ?

### ISSUE THREAD

**Issue by Kochise** [rvaiya/keyd#634]
Hi,

looking for a `Linux` keyboard layout remapper to convert a `msklc` of mine, if possible including media keys as well.

I'm reading https://github.com/rvaiya/keyd/blob/master/docs/keyd.scdoc but it's cryptic to say the least.

Basically, my need is to convert some keys/scan code (not **ASCII** character that may move around regarding a specific firmware mapping that may be existing below `keyd`) to a unicode character.

That includes normal key press (1), Shift (2), AltGr (3, aka Ctrl+Alt) and AltGr+Shift (4, aka Ctrl+Alt+Shift).

My custom keyboard layout includes a lot of "dead keys" to output accented characters. `msklc` requests to provide a list of all possible combinations/output values after a deadkey has been pressed. How to perform this with `keyd` ?

Regards.

**Comment by nsbgn** [comment]
> I'm reading https://github.com/rvaiya/keyd/blob/master/docs/keyd.scdoc but it's cryptic to say the least.

What would have helped you grok it quicker? In short, `keyd` uses the compose key to emit non-ASCII characters. This is a dead key used by your display server --- if so instructed. To activate it, use `setxkbmap -option compose:menu` for X11 (or something compositor-dependent on Wayland). You can now also use stock compose sequences, independent of keyd, which I find usually preferable --- eg `<compose> <'> <a>` will output `á`.

Otherwise, you need to `ln -s /usr/share/keyd/keyd.compose ~/.XCompose` so that the compose sequence emitted by `keyd` is actually understood by your display server. You can then simulate custom dead keys by using `oneshot()`. For example:

```
[main]
' = oneshot(deadkey_apostrophe)
# ...

[deadkey_apostrophe]
' = '
a = á
e = é
# ...
```

> That includes normal key press (1), Shift (2), AltGr (3, aka Ctrl+Alt) and AltGr+Shift (4, aka Ctrl+Alt+Shift).

You can use composite layers for combinations:

```
[deadkey_apostrophe+shift]
A = Á
E = É
# ...
```

**Comment by Kochise** [comment]
Ok, thank for your reply, hence I'll go your way, but again it's not straightforward, having to edit this and tweak there...

My `msklc` file is here : https://github.com/wenuam/wm_key_layout_cfilorux/blob/main/map/Windows/msklc/cfilorux.klc

The installer is provided if you want to give it a try.

**Comment by Kochise** [comment]
@nsbgn Here's what I did, but it doesn't seems to work, it doesn't even reload.

Could you advice me how to fix it ?

[default.conf.txt](https://github.com/rvaiya/keyd/files/13780837/default.conf.txt) (rename to .conf)

**Comment by Kochise** [comment]
Ok, after further investigation it appears that `keyd` is not suitable, I'll then try another contenders.

Thank anyway.

**Comment by nsbgn** [comment]
>again it's not straightforward, having to edit this and tweak there

Nothing is straightforward once you go into custom keyboard land :P

> Could you advice me how to fix it ?

Probably run the `keyd` daemon itself first (`systemctl start keyd` for systemd, or just `keyd` to run it manually).

> it appears that keyd is not suitable

Yeah, I was going to say that for this specific use case, you might be better off writing your own `.Xcompose` file that corresponds to whatever sequences are defined in the original file. Good luck.

**Comment by Kochise** [comment]
I did everything as requested, yet when I do `sudo keyd reload` and get via `systemctl status keyd` something like :

`keyd[750]: keyd: src/ini.c:109: ini_parse_string: Assertion 'n < MAX_SECTIONS' failed.`

I consider this a big flashing red light and an absolute no-go for something that should be more trivial to use in 2023/2024, custom keyboard land or not (`msklc` made it quite easy for me, and is more than 20 yo already).

I just cannot expect anyone to tweak into the source code to remove artificial limitations, while `MAX_SECTION_ENTRIES` set at 1024 can covers more than 10 keyboards worth of remapping.

No wonder `Linux` isn't yet ready for wider acceptance. This thing is a hack. And hard to tweak.

**Comment by nsbgn** [comment]
It's perfectly fine to conclude that keyd doesn't cover your use case well (I agree!), or to point out issues, or to discover that you're not the target audience. FOSS openness means choice and community and autonomy, so you can use one of the many other projects, or help to improve this one, or fork it, or make something new. But pouting that a niche, gratis, volunteer-based project casts a dark shadow on all of Linux because it happens to not immediately satisfy your particular consumer demands is, frankly, rude. It also contributes to open source maintainer burnout. Please consider that when engaging with FOSS projects in the future.

**Comment by Kochise** [comment]
Sure I do, thank for supporting `keyd`, but "niche, gratis, volunteer-based project" can still be a bit more professional instead to cocoon itself into self-sufficiency and request to the users, not the developers, to have to get their hands dirty.

It's like DIY car makers wanting to be like regular manufacturers, yet not taking examples on usability and ergonomic. Then wonders why people are not wanting to pour as much time as they did to "create their own car" when all those people wants is sure a "customizable car", but not having to build the engine themselves or in my case, the car's steering wheel.

It's a global mindset, wanting to be a top player ("look, Linux is everywhere, so it must be of value") yet not wanting to round the rough edges.

For instance, those artificial limitations are probably due to static array reservation, using a defined count of cells, yet the C language allows malloc/free for quite a long time now, thus trying to diminish the importance of such trivial architectural choice is not a solid argument to me, just a lack of foresight, pretty much like the famous "640KB should be enough for everyone".

I'd like to participate to many projects out there, I did it already and continue to do so, but all projects have different languages, styles, motivations, it's hard to follow track. Hence sometimes I'm just on the "user" side and wanting to use a carefully crafted piece of code, promoted with bells and whistles, only to find out it's not as easy/useful/configurable/... as I thought.

So, I guess the problem is all mine, I probably haven't read the doc enough (which doesn't state anything about those hard-coded limitations) and should be more vigilant. But again, how are you supposed to select the right candidate before trying them, without guarantee it will work, considering the various ways to perform it, the different compositors, distros, etc. 

Anyway, Happy New Year, keep it strong.

Regards.

**Comment by nsbgn** [comment]
We seem to have wildly incompatible standards of professionalism. You also seem to operate under the belief that Linux software is only ever written as a means of capturing mainstream market share for Linux.

So as not to veer off-topic, I'll otherwise only address your technical complaint:

> For instance, those artificial limitations are probably due to static array reservation, using a defined count of cells, yet the C language allows malloc/free for quite a long time now, thus trying to diminish the importance of such trivial architectural choice is not a solid argument to me, just a lack of foresight, pretty much like the famous "640KB should be enough for everyone".

rvaiya has historically been pretty receptive to requests for increasing limits (or making them dynamic) if you can make a good argument, but I don't think that the assertion that it's trivial constitutes one. I agree with you that keyd (currently) isn't the best solution for hacking your way to thousands of dead keys, so this is hardly a compelling reason to change the limit.

What initially attracted me to keyd was precisely its conservative approach, as described in the README. Having the 'foresight' to address unimagined use cases often means making concessions in other areas. That isn't necessarily bad, and might well fit another project — but I appreciate that keyd is carefully crafted around a limited scope that is expanded only after deliberating with users.

**Comment by Kochise** [comment]
Fair to me :)
